src/processor.rs

```diff
@@ -178,52 +199,75 @@ impl Processor {
+                    select! {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(5)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = stats_publisher.publish_stats(redis.clone()).await {
                         error!("Error publishing processor stats: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop web metrics");
             }
-        });
+        }));
 
         // Start retry and scheduled routines.
-        tokio::spawn({
+        handles.push(tokio::spawn({
             let redis = self.redis.clone();
+            let cancellation_token = self.cancellation_token.clone();
             async move {
                 let sched = Scheduled::new(redis);
                 let sorted_sets = vec!["retry".to_string(), "schedule".to_string()];
 
                 loop {
                     // TODO: Use process count to meet a 5 second avg.
-                    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+                    select! {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(5)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = sched.enqueue_jobs(chrono::Utc::now(), &sorted_sets).await {
                         error!("Error in scheduled poller routine: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop for retry and scheduled");
             }
-        });
+        }));
 
         // Watch for periodic jobs and enqueue jobs.
-        tokio::spawn({
+        handles.push(tokio::spawn({
             let redis = self.redis.clone();
+            let cancellation_token = self.cancellation_token.clone();
             async move {
                 let sched = Scheduled::new(redis);
 
                 loop {
                     // TODO: Use process count to meet a 30 second avg.
-                    tokio::time::sleep(std::time::Duration::from_secs(30)).await;
+                    select! {
+                        _ = tokio::time::sleep(std::time::Duration::from_secs(30)) => {}
+                        _ = cancellation_token.cancelled() => {
+                            break;
+                        }
+                    }
 
                     if let Err(err) = sched.enqueue_periodic_jobs(chrono::Utc::now()).await {
                         error!("Error in periodic job poller routine: {:?}", err);
                     }
                 }
+
+                debug!("Broke out of loop for periodic");
             }
-        });
+        }));
 
-        loop {
-            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        for handle in handles {
```

Comment:
I'm a big fan of this. I never liked `tokio::spawn` calls without a good way to clean up.